### ISSUE TITLE
HGL: Hierarchical Geometry Learning for Test-time Adaptation in 3D Point Cloud Segmentation

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#58]
<img width="1212" alt="image" src="https://github.com/user-attachments/assets/18b2ae52-a7a6-4bdd-875f-1ee632ccfa30">

### Code Link
https://github.com/tpzou/HGL

## Task Definition
**Test-time adaptation** (TTA): model adaptation is performed by learning from the unlabeled test (target) data during inference time.

**3D point cloud semantic segmentation** is to assign a label to each point in the point cloud.

**TTA in 3D point cloud segmentation (TTA3DSeg)**, a pre-trained 3D segmentation model to be updated online to a new target domain without access to the source domain.

## Motivation

Threefold motivation for parsing continuous unstructured point cloud streams into **local** (point-level), **global** (object-level), and **temporal** (frame-level) hierarchical structures:

at the **point** level, each individual point exhibits _semantic attributes highly correlated with its neighboring points_; 

at the **object** level, despite distribution shifts, objects of the _same category_ should manifest _similar feature embeddings_; and 

at the **frame** level, the _embedding features_ of geometrically corresponding points _from different frames_ should demonstrate _similarity_.

### Local Geometry Learning

Existing methods typically generate pseudo-labels directly based on **single-point prediction**. Despite reasonable effectiveness, these methods neglect the local geometry structure, leading to sub-optimal results.

To tackle this, we first **aggregate the predictions** from spatial neighbors to generate more accurate pseudo-labels. Then, we introduce **geometric purity** and **prediction certainty confidence** to select reliable local geometric pseudo-labels.

<img width="600" alt="image" src="https://github.com/user-attachments/assets/ea5b6559-4f88-4848-9399-638096598c1e">

<img width="600" alt="image" src="https://github.com/user-attachments/assets/ed225279-6dd1-42f2-9614-fa395d8c8184">

<img width="600" alt="image" src="https://github.com/user-attachments/assets/ee599026-ff43-4b6b-8c8a-fa6fbcb48fa2">

https://github.com/tpzou/HGL/blob/60f2a6a59011feb4bb1766eccad8e92f444049ab/pipelines/adaptation_online_single.py#L586

<img width="600" alt="image" src="https://github.com/user-attachments/assets/90f39f2d-bb06-4fc0-b3a8-1f7e83d4fb3d">

Consequently, at time $t$ and for class $c$, we exclusively choose local geometric pseudo-labels with confidence scores higher than $a$, utilizing them as selected local geometric pseudo-labels.

### Global Geometry Fine-tuning

1. The local geometric pseudo-labels are obtained from the source model only. Therefore, the generated pseudo-labels are still **noisy due to domain shift** and cannot benefit from the continuously increasing target domain data.
2. The selected local geometric pseudo-labels are chosen by λ. It is **challenging to ensure a good balance** between accuracy and generalization in the selected samples.

Hence, we need to fine-tune local geometric pseudo-labels by calculating the **global affinity between the global target prototypes and the feature of each point.**

**Class Prototype Calculation**

<img width="600" alt="image" src="https://github.com/user-attachments/assets/d10c7e0a-3dca-416a-a925-93888f3a9ae5">

**Update Prototype via EMA and Compute Global Geometric Pseudo-label**

<img width="600" alt="image" src="https://github.com/user-attachments/assets/ab26fc02-6b6e-4f79-b3fe-5eac0ac5fed8">

**Fine-tune the Local Geometric Pseudo-labels**
<img width="600" alt="image" src="https://github.com/user-attachments/assets/e97a38a3-a70b-4868-ab7d-4c958b4fa27f">

<img width="700" alt="image" src="https://github.com/user-attachments/assets/c245bad4-70c3-495d-83e0-57a7e25ca7a5">

### Temporal Geometry Regularization

<img width="500" alt="image" src="https://github.com/user-attachments/assets/d60a5cb0-cfb1-4f99-8e91-159dbb8f5e76">

Considering the **temporal geometric consistency** in local-global pseudo-labeling-based self-supervised learning is effective in smoothing the prediction and preventing negative transfer.

<img width="600" alt="image" src="https://github.com/user-attachments/assets/02d03e64-2ce5-4a81-9fe5-5739122ca630">

https://github.com/tpzou/HGL/blob/60f2a6a59011feb4bb1766eccad8e92f444049ab/pipelines/adaptation_online_single.py#L863

## Experiments 

<img width="979" alt="image" src="https://github.com/user-attachments/assets/c8dee918-0cac-4789-9a36-27836ee1af85">

<img width="996" alt="image" src="https://github.com/user-attachments/assets/036beea8-66a2-403b-8854-c90567d89fa4">

<img width="1003" alt="image" src="https://github.com/user-attachments/assets/2eacc47d-d58b-4a1c-9752-174a38a489db">

<img width="643" alt="image" src="https://github.com/user-attachments/assets/6958ba4f-0cc3-4835-a173-ec5a3c8b25eb">

<img width="1312" alt="image" src="https://github.com/user-attachments/assets/49fce2fa-432a-4cb0-9cc6-b7f0e6211087">

<img width="925" alt="image" src="https://github.com/user-attachments/assets/fa130da6-a50b-4044-90af-a77058c17a9c">

## Notes

A lot of errors in equations. 

The methods presented in this paper are straightforward and integrate many prior works. This paper's acceptance to ECCV is attributed to its status as an early attempt at TTA3DSeg (the second such attempt), the extensive experiments conducted, the excellent performance achieved, and the satisfactory quality of the writing.
